### ISSUE TITLE
Is Required validation fails

### ISSUE THREAD

**Issue by GoogleCodeExporter** [viktorthang/wordpress-custom-content-type-manager#374]
```
I have 2 custom text fields. I want to be able to put in hex color values. I 
also have validation on the fields making them required. If the user enters 
000000 as a value the validation fails and says the field must be entered. The 
value can be entered in quotes "000000", but when they go back in it shows 0 as 
the value and they can not save. 000007 is saved as 7. I want this to be 
treated as a regular text field allowing the user to enter anything they want 
and to validate if they have a value or the field is empty. I don't want it to 
change I want it to remain exactly as it was entered. 
```

Original issue reported on code.google.com by `ira...@gmail.com` on 18 Apr 2012 at 3:36

**Comment by GoogleCodeExporter** [comment]
```
Changed fields to text area and the same thing is happening .
```

Original comment by `ira...@gmail.com` on 18 Apr 2012 at 3:43

**Comment by GoogleCodeExporter** [comment]
```
Interesting.  I think this is a PHP bug because "000000" in any flavor 
evaluates as empty.  In PHP wackiness:

"foo" == TRUE, and "foo" == 0… but, of course, TRUE != 0

Let me re-evaluate the approach here... maybe strlen might help.  Thank you so 
much for reporting this use case!

```

Original comment by `ever...@fireproofsocks.com` on 18 Apr 2012 at 6:05
- Changed state: **InReview**

**Comment by GoogleCodeExporter** [comment]
```
Curiously, the problem lies in the JSON decode.  Inside CCTM_FormElement.php, 
the get_value() function around line 600:

$out = (array) json_decode($str, true );

if (empty($out)) {
    return $str;
}
else {
    return $out[0];
}


JSON decoding does not like leading zeros!  e.g. 

00001 decodes to: Array ( [0] => 1 )
```

Original comment by `ever...@fireproofsocks.com` on 18 Apr 2012 at 6:58
- Changed title: **Is Required validation fails**
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
im having them evter the hex codes with # at the beginning and that seems to 
work.
```

Original comment by `ira...@gmail.com` on 18 Apr 2012 at 9:44

**Comment by GoogleCodeExporter** [comment]
```
That will work because jsone_decode will interpret #0000 as a string (whereas 
0000 gets interpreted as numeric).  The fix is this:

Inside includes/CCTM_FormElement.php, incise the get_value() function, add the 
following around line 600:


            if (is_numeric($str)) {
                return $str;
            }


That should come immediately after the lines containing this:

            if ($str=='[""]') {
                return '';
            }

This will be on the dev branch shortly.
```

Original comment by `ever...@fireproofsocks.com` on 19 Apr 2012 at 6:11
- Changed state: **Fixed**
- Added labels: **Milestone-Release0.9.6**

**Comment by GoogleCodeExporter** [comment]
```
See 
http://wordpress.org/support/topic/value-of-0-in-a-required-text-field?replies=7
#post-4167745

0 should be a valid response for a required field.  I.e. fall back to strlen > 
0.
```

Original comment by `ever...@fireproofsocks.com` on 6 May 2013 at 11:15
- Changed state: **ReOpened**

**Comment by GoogleCodeExporter** [comment]
```
In includes/StandardizedCustomFields.php, the block around line 111 should use 
strlen instead of the empty.  Here's the revised block:

if ($FieldObj->required) {
    if ((is_array($value_copy) && !empty($value_copy))
        || (!is_array($value_copy) && !strlen(trim($value_copy)))) {
        CCTM::$post_validation_errors[$FieldObj->name] = sprintf(__('The %s field is required.', CCTM_TXTDOMAIN), $FieldObj->label);
    }
}

Committed revision 708950.
```

Original comment by `ever...@fireproofsocks.com` on 6 May 2013 at 11:46
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
The fix included in version 0.9.7.3 does not seem to have resolved allowing a 0 
as a value; it no longer gives any error messages, but it will not allow the 
post to be published.  Attempting to publish results in a draft.
```

Original comment by `mfrec...@gmail.com` on 14 May 2013 at 10:01

**Comment by GoogleCodeExporter** [comment]
```
Ack... !#$%!#$^ subtle thing.  Edit line 517 in StandardizedCustomFields.php 
and change to this:

if ($FieldObj->required && !strlen(trim($value_copy))) {
```

Original comment by `ever...@fireproofsocks.com` on 14 May 2013 at 10:28
